identity/src/main/java/com/android/identity/credential/Credential.kt

```diff
@@ -0,0 +1,330 @@
+     */
+    val replacementFor: Credential?
+        get() = document.certifiedCredentials.firstOrNull { it.identifier == replacementForIdentifier }
+            .also {
+                if (it == null && replacementForIdentifier != null) {
+                    Logger.w(
+                        TAG, "Credential with alias $replacementForIdentifier which " +
+                                "is intended to be replaced does not exist"
+                    )
+                }
+            }
+
+    /**
+     * The credential that will replace this credential once certified or `null` if no
+     * credential is designated to replace this credential.
+     */
+    val replacement: Credential?
+        get() = document.pendingCredentials.firstOrNull { it.identifier == replacementIdentifier }
+            .also {
+                if (it == null && replacementIdentifier != null) {
+                    Logger.w(
+                        TAG, "Pending credential with identifier $replacementIdentifier which " +
+                                "is intended to replace this credential does not exist"
+                    )
+                }
+            }
+
+    /**
+     * Deletes the credential.
+     *
+     * After deletion, this object should no longer be used.
+     */
+    open fun delete() {
+        document.removeCredential(this)
+    }
+
+    /**
+     * Increases usage count of the credential.
+     */
+    fun increaseUsageCount() {
+        usageCount += 1
+        document.saveDocument()
+    }
+
+    /**
+     * Certifies the credential.
+     *
+     * @param issuerProvidedAuthenticationData the issuer-provided static authentication data.
+     * @param validFrom the point in time before which the data is not valid.
+     * @param validUntil the point in time after which the data is not valid.
+     */
+    fun certify(
+        issuerProvidedAuthenticationData: ByteArray,
+        validFrom: Timestamp,
+        validUntil: Timestamp
+    ) {
+        check(!isCertified) { "Credential is already certified" }
+        isCertified = true
+        _issuerProvidedData = issuerProvidedAuthenticationData
+        _validFrom = validFrom
+        _validUntil = validUntil
+
+        replacementFor?.delete()
+        replacementForIdentifier = null
+
+        document.certifyPendingCredential(this)
+    }
+
+    /**
+     * Utility function for when a credential is created for the first time. Called by [Credential]
+     * constructor and constructors of inheriting classes.
+     *
+     * @param document the document associated with the credential.
+     * @param asReplacementFor the credential this credential will replace, if not null
+     * @param domain the domain of the credential
+     * @return the credential object.
+     */
+    protected fun create(
+        document: Document,
+        asReplacementFor: Credential?,
+        domain: String
+    ): Credential {
```

Comment:
I think it would be nice if we can avoid passing a `Document` to the any of the constructors... instead, have `Document.addCredential()` set the `Document` instance (via an internal method on `Credential`).